spec/cov/cov_spec.rb

```diff
@@ -0,0 +1,177 @@
+# frozen_string_literal: true
+
+require "datadog_cov.#{RUBY_VERSION}_#{RUBY_PLATFORM}" unless PlatformHelpers.jruby?
+
+require_relative "calculator/calculator"
+require_relative "calculator/code_with_❤️"
+
+RSpec.describe "Datadog::CI::Cov" do
+  before do
+    skip "Code coverage is not supported in JRuby" if PlatformHelpers.jruby?
+  end
+
+  def absolute_path(path)
+    File.expand_path(File.join(__dir__, path))
+  end
+
+  subject { Datadog::CI::Cov.new(root: root) }
+
+  describe "code coverage collection" do
+    let!(:calculator) { Calculator.new }
+    context "when root is the calculator project dir" do
+      let(:root) { absolute_path("calculator") }
+
+      it "collects code coverage including Calculator and operations" do
+        subject.start
+
+        expect(calculator.add(1, 2)).to eq(3)
+        expect(calculator.subtract(1, 2)).to eq(-1)
+
+        coverage = subject.stop
+
+        expect(coverage.size).to eq(3)
+        expect(coverage.keys).to include(
+          absolute_path("calculator/calculator.rb"),
+          absolute_path("calculator/operations/add.rb"),
+          absolute_path("calculator/operations/subtract.rb")
+        )
+      end
+
+      it "does not support files with non-ASCII characters yet due to additional overhead of UTF-8 strings parsing" do
+        subject.start
+        expect(I❤️Ruby.new.call).to eq("I ❤️ Ruby")
+        coverage = subject.stop
+        expect(coverage.size).to eq(1)
+        expect(coverage.keys.first).to include("calculator/code_with_")
+      end
```

Comment:
yes, it is broken, there are a bunch of UTF-8 codepoints there

I tried to use `rb_enc_str_new` but as expected it increased the overhead: I will get back to it as soon as I start optimizing code coverage collection. My goal is to fix it eventually or provide config to use UTF-8 encoding for filenames as opt-in

Reply:
Hmm... Indeed with `rb_sourceline()` you may be slightly stuck between a rock and a hard place.

You could do the encoding lazily, by e.g. keeping the bytes in the map, and only resolving them later.

Alternatively, it may be worth experimenting with something other than `rb_sourceline`. `rb_profile_frames` can be configured to only give you the top of the stack, and then you can ask the file for it and get a `VALUE` directly.

There's also `rb_debug_inspector_backtrace_locations` and `rb_make_backtrace` although I haven't used them before and suspect they may be more expensive.

---

One tip for this test is that the expectations right now don't seem to state very clearly that the string will be broken. Maybe worth adding a comment or something that states very clearly "this string will be incorrect" -- right now if you tweak the test description to say "it supports files with non-ascii characters" the test still seems correct ;)